skimage/feature/corner.py

```diff
@@ -956,38 +976,53 @@ def corner_peaks(image, min_distance=1, threshold_abs=None, threshold_rel=None,
            [2, 3],
            [2, 2]])
     >>> corner_peaks(response, threshold_rel=0)
-    array([[2, 2]])
+    array([[3, 3]])
 
     """
     if threshold_rel is None:
         threshold_rel = 0.1
         warn("Until version 0.16, threshold_rel was set to 0.1 by default. "
              "Starting from version 0.16, the default value is set to None. "
              "Until version 0.18, a None value corresponds to a threshold "
-             "value of 0.1. "
-             "The default behavior will match skimage.feature.peak_local_max. "
-             "To avoid this warning, set threshold_rel=0.",
-             category=FutureWarning, stacklevel=2)
-
-    peaks = peak_local_max(image, min_distance=min_distance,
-                           threshold_abs=threshold_abs,
-                           threshold_rel=threshold_rel,
-                           exclude_border=exclude_border,
-                           indices=False, num_peaks=num_peaks,
-                           footprint=footprint, labels=labels,
-                           num_peaks_per_label=num_peaks_per_label)
-    if min_distance > 0:
-        coords = np.transpose(peaks.nonzero())
-        for r, c in coords:
-            if peaks[r, c]:
-                peaks[max((r - min_distance), 0):r + min_distance + 1,
-                      max((c - min_distance), 0):c + min_distance + 1] = False
-                peaks[r, c] = True
-
-    if indices is True:
-        return np.transpose(peaks.nonzero())
-    else:
-        return peaks
+             "value of 0.1. The default behavior will match "
+             "skimage.feature.peak_local_max.", category=FutureWarning,
+             stacklevel=2)
+
+    if np.isinf(num_peaks):
+        num_peaks = None
+
+    # Get the coordinates of the detected peaks
+    coords = peak_local_max(image, min_distance=min_distance,
+                            threshold_abs=threshold_abs,
+                            threshold_rel=threshold_rel,
+                            exclude_border=exclude_border,
+                            indices=True, num_peaks=np.inf,
+                            footprint=footprint, labels=labels,
+                            num_peaks_per_label=num_peaks_per_label)
+
+    if len(coords):
+        # Use KDtree to find the peaks that are too close to each other
+        tree = spatial.cKDTree(coords)
```

Comment:
@rfezzani  Have you played around with the `balanced_tree=True` parameter? Constructing a balanced tree takes more time but should pay of the more we queries we use on it (which should be quite a lot in this case).

Reply:
I did not, but I definitely will. I will come back after some more tests ;-)

Reply:
@lagru, I made some more investigation, and I observed no influence of the `balanced_tree` parameter on the computation time:

```python
In [1]: from skimage.feature import corner_peaks, corner_fast                                                                           

In [2]: from skimage.color import rgb2gray                                                                                              

In [3]: from skimage.data import astronaut                                                                                              

In [4]: img =rgb2gray(astronaut())                                                                                                      

In [5]: corners = corner_fast(img, 12, 0.3)                                                                                             

In [6]: %timeit corner_peaks(corners, threshold_rel=0.1) 
```

#### `balanced_tree=True`

```python                                                                
11.6 ms ± 20.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

#### `balanced_tree=False`

```python                                                               
11.5 ms ± 49.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Reply:
Thanks for investigating!